basic_pipeline/8_Mixer.md

```diff
@@ -0,0 +1,232 @@
+  {[buffer | buffers], tracks}
+  else
+    {[], tracks}
+  end
+ end
+ ...
+end
+```
+
+In order to output the buffers, we need to fetch the desired buffers - that is what we do with the `prepare_buffers/1` function. Later on, we are simply creating the `:buffer` action, basing on the list of buffers to be output. 
+In the `prepare_bufers/1` we get all the active tracks (by 'active' we mean that there is still an unprocessed buffer in the `Track` structure - independent of the status of that track). If all the active tracks have the buffers we can output the one with the lowest presentation timestamp and recursively call the `prepare_buffers/1` (in case there are some buffers that still need to be output - this can happen in a 'corner case' of processing the buffer from the track in the `:finished` state). Surely, we also need to update the state so that to remove the processed buffers.
+Now let's focus on preparing `:end_of_stream` action:
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp send_end_of_stream(state) do
+  end_of_stream_actions =
+    if Enum.all?(state.tracks, fn {_, track} -> track.status == :finished end) do
+      [end_of_stream: :output]
+    else
+      []
+    end
+ {state, end_of_stream_actions}
+ end
+ ...
+end
+```
+This action needs to be sent if both the tracks are in the `:finished` state - since the `send_end_of_stream/1` function gets called after the `output_buffers/1`, we can be sure, that all the buffers which could possibly be on those tracks, despite they are in the `:finished` state, are already processed.
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp demand_on_empty_tracks(state, pads) do
+  actions =
+    state.tracks
+    |> Enum.filter(fn {track_id, track} ->
+      track.status != :finished and track.buffer == nil and pads[track_id].demand == 0
+    end)
+    |> Enum.map(fn {track_id, _} -> {:demand, {Pad.ref(track_id), 1}} end)
+
+  {state, actions}
+ end
+ ...
+end
+```
+The last type of actions we need to generate are`:demand` actions. From the [context](https://hexdocs.pm/membrane_core/Membrane.Element.CallbackContext.Demand.html#t:t/0) passed as one of the arguments in the `handle_demand/5` callback, we have passed the `context.pads`.
+That is how we can fetch the information about the current demand size on the given pad.
+For all the tracks which are not yet `:finished`, do not have the buffer and the demand was not made on behalf of that pad (there is where we are making usage of the context information - `pads[track_id].demand==0`), we are making such a demand for one buffer.
+
+Starting from that moment, our mixer should be capable of mixing the inputs from two sources! In the later part of this tutorial, we will extend the mixer so that it will be able to mix any number of tracks we can desire (and we will be able to change that number during the pipeline runtime).
```

Comment:
```suggestion
Starting from that moment, our mixer should be capable of mixing the inputs from two sources! In the later part of this tutorial, we will extend the mixer so that it will be able to mix any number of tracks.
```

Reply:
done